### ISSUE TITLE
Underscore in kong.response headers

### ISSUE THREAD

**Issue by Asafb26** [Kong/kong#6995]
### Summary

It looks like kong transform underscore into dash while setting a new response header

### Steps To Reproduce
```lua
function Test:header_filter(conf)
  kong.response.set_header('service_name', 'test')
end
```

results with `service-name`

### Additional Details & Logs

- Kong version - 2.3.2

**Comment by kikito** [comment]
Thanks for reporting this. We're investigating this problem.

**Comment by bungle** [comment]
There is this option:
https://github.com/openresty/lua-nginx-module#lua_transform_underscores_in_response_headers

```
KONG_NGINX_HTTP_LUA_TRANSFOR_UNDERSCORES_IN_RESPONSE_HEADERS=off
```

**Comment by bungle** [comment]
> (but I am not sure if it breaks anything)

Prometheus plugin for example has code like this:
```
ngx.header.content_type = "text/plain"
```

**Comment by kikito** [comment]
Hello @Asafb26, as @bungle is pointing out, some plugins do rely on the automatic conversion of underscores into slashes, so we can't "fix" this without breaking someone else's code.

A possible workaround is just setting `lua_transform_underscores_in_response_headers` to `off`. The tradeoff is that doing so might "break" plugins that rely on the underscore transformation, like Prometheus (I don't think this would "break" the Prometheus plugin, but it might make it miss the Content-Type header on some responses).

Palliative mesures:

* I am documenting this behavior officially (plus workaround) in #7018 
* I have opened a PR to the Prometheus Plugin (https://github.com/Kong/kong-plugin-prometheus/pull/124) to stop using underscores, so it is fully impervious to `lua_transform_underscores_in_response_headers`
* I have reviewed all the bundled plugins and the Kong source itself. I did not find another place in which we used underscored header names.

**Comment by Asafb26** [comment]
Hi @kikito , Thanks 🙏
I guess that we may change this behavior in the near future? as a Prometheus user, I'm thinking that toggle off the `lua_transform_underscores_in_response_headers` can be lead to some issues

**Comment by kikito** [comment]
> I guess that we may change this behavior in the near future? as a Prometheus user, I'm thinking that toggle off the lua_transform_underscores_in_response_headers can be lead to some issues

The change in the prometheus plugin seems fairly small. The way I see it your options are:

* Just wait until we release a new version of the Prometheus plugin (this can happen independently from Kong releases). Then you can use that version in your Kong rockspec and activate `lua_transorm_underscores_in_response_headers`
* Manually modify the code of the Prometheus plugin yourself. How to implement this depends greatly on your infrastructure setup. The patch to apply is very small (just two lines) but the files can be placed in lots of locations depending on Linux distros etc.
* Or you could fork the plugin, release your own version of the rockspec to luarocks, and use that while we release ours.

I would use option 1 if I could wait and option 3 if I couldn't wait.

For now, since this issue is "addressed" in 2.4, I will close the issue.
